src/coreclr/src/vm/processdiagnosticsprotocolhelper.cpp

```diff
@@ -0,0 +1,214 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "fastserializer.h"
+#include "processdiagnosticsprotocolhelper.h"
+#include "eventpipeeventsource.h"
+#include "diagnosticsprotocol.h"
+#include "diagnosticserver.h"
+
+#ifdef FEATURE_PERFTRACING
+
+static inline uint32_t GetStringLength(const WCHAR *&value)
+{
+    return static_cast<uint32_t>(wcslen(value) + 1);
+}
+
+static bool TryWriteString(uint8_t * &bufferCursor, uint16_t &bufferLen, const WCHAR *&value)
+{
+    uint32_t stringLen = GetStringLength(value);
+    S_UINT16 totalStringSizeInBytes(stringLen * sizeof(WCHAR) + sizeof(uint32_t));
+    ASSERT(!totalStringSizeInBytes.IsOverflow());
+    ASSERT(bufferLen >= totalStringSizeInBytes.Value());
+    if (bufferLen < totalStringSizeInBytes.Value() || totalStringSizeInBytes.IsOverflow())
+        return false;
+
+    memcpy(bufferCursor, &stringLen, sizeof(stringLen));
+    bufferCursor += sizeof(stringLen);
+
+    memcpy(bufferCursor, value, stringLen * sizeof(WCHAR));
+    bufferCursor += stringLen * sizeof(WCHAR);
+
+    bufferLen -= totalStringSizeInBytes.Value();
+    return true;
+}
+
+uint16_t ProcessInfoPayload::GetSize()
+{
+    LIMITED_METHOD_CONTRACT;
+
+    // The protocol buffer is defined as:
+    // X, Y, Z means encode bytes for X followed by bytes for Y followed by bytes for Z
+    // uint = 4 little endian bytes
+    // long = 8 little endian bytes
+    // GUID = 16 little endian bytes
+    // wchar = 2 little endian bytes, UTF16 encoding
+    // array<T> = uint length, length # of Ts
+    // string = (array<wchar> where the last char must = 0) or (length = 0)
+
+    // uint64_t ProcessId;
+    // GUID RuntimeCookie;
+    // LPCWSTR CommandLine;
```

Comment:
The spec should clarify all the expected formatting rules for command line strings and enumerate all string constants for OS and Arch.

Reply:
Let me spin up a PR over in dotnet/diagnostics.  I'll update the comments here to link to that which should reflect the updated spec once it gets merged.